### ISSUE TITLE
“eval” can delete variables from namespaces

### ISSUE THREAD

**Issue by hanche** [elves/elvish#1492]
Here is how I can use `eval` to delete a variable from an existing namespace:

```
⬥ var tt: = (ns [&foo=bar])
⬥ echo $tt:foo
bar
⬥ eval 'del foo' &ns=$tt:
⬥ keys $tt:
⮕ foo
⬥ echo $tt:foo
Exception: variable $tt:foo not found
[tty 5], line 1: echo $tt:foo
⬥ echo $tt:[foo]
Exception: no such key: foo
[tty 6], line 1: echo $tt:[foo]
```

This does not seem desirable. Especially the inconsistency, with the variable still listed as a key of the namespace.

This trick also lets me delete a variable from a module namespace introduced by `use`.

**Comment by krader1961** [comment]
Ugh! Unsurprisingly this can be used to delete variables defined in builtin namespaces but it also no longer shows up in the output of `keys`:

```
> keys $edit: | to-lines | grep command-duration
command-duration
> echo $edit:command-duration
0.000345857
> eval 'del command-duration' &ns=$edit:
> echo $edit:command-duration
Exception: variable $edit:command-duration not found
> keys $edit: | to-lines | grep command-duration
Exception: grep exited with 1
```

**Comment by krader1961** [comment]
Note that fixing this implies that @hanche's attempt to use `eval` as a substitute for `use` won't work. This example by @hanche on IRC earlier today will stop working if [`eval`](https://elv.sh/ref/builtin.html#eval) is changed to conform to its documentation with respect to "namespace specified by &ns is never modified; it will not be affected by the creation or deletion of variables by $code":

```
var bar:
eval (slurp <foo.elv) &ns=(ns [&]) &on-end={|ns| set bar: = $ns }
```

**Comment by hanche** [comment]
@krader1961 I don't think that's right. Let me explain why.

But first: I incidentally discovered that _if the eval'd code creates a new variable_, a `del` does _not_ affect the original namespace! Here is a working example, which acts just like I had expected it to:

```
⬥ var orig: = (ns [&x=xxx &y=yyy])
⬥ eval 'del x; set y = new; var z = zzz' &ns=$orig: &on-end={|ns| set new: = $ns }
⬥ keys $orig:
⮕ x
⮕ y
⬥ put $orig:x $orig:y
⮕ xxx
⮕ new
⬥ keys $new:
⮕ y
⮕ z
⬥ put $new:y $new:z
⮕ new
⮕ zzz
```

Indeed, the way I thought it was supposed to work, and which the above experiment seems to confirm, was as follows:

1. The eval'd code operates not exactly in the provided namespace, but in a _clone_ of that namespace.
2. Hence any alterations to that namespace will not be seen in the original (but altering the _value_ of a variable will produce visible effects in the original)
3. Finally, the altered namespace is provided as an argument to the `on-end` function

And this is exactly how I imagine the loading of a module works, except the initial provided namespace is always empty. And instead of the `on-end` mechanism, the language caches the namespace and assigns it to a suitable variable in the namespace of the caller. In particular, my “`eval` trick” should still work.

I think today's discovery offers some insight into what is happening behind the curtain, but I won't speculate as to exactly what is going on.

**Comment by krader1961** [comment]
> But first: I incidentally discovered that if the eval'd code creates a new variable, a del does not affect the original namespace! 

Yes, that is my point. The documentation (as I write this) states that a namespace cannot be modified by adding or deleting a variable in the namespace. That is not true. A namespace can be modified but doing so creates a new namespace. Which is (or should be) only useful in the context of something like `eval` where you can capture the new namespace. My prior comment(s) were meant to point out that the documentation is misleading or the current implementation of namespaces does not conform to the documentation.
